python/mxnet/pretrained_model.py

```diff
@@ -0,0 +1,576 @@
+        ret['rank'] = ranks
+        return ret
+
+class ImageDetector(object):
+    """
+    Wrapeer of pretrained image detect model.
+
+    Use :py:func:`load_model` or :py:func:`load_path` to load the model. Do not
+    construct directly.
+
+    Parameters
+    ----------
+    model: FeedForward
+        The underlying model
+    labels : list
+        Map from index to label
+    metadata: dict
+        Metadata of the model including name, version, input shape, etc
+    """
+    def __init__(self, model, labels, metadata):
+        self._model = model
+        self._labels = labels
+        self._metadata = metadata
+        self._num_classes = len(labels)
+        # rcnn params
+        for k, v in metadata.items():
+            self.__setattr__("_" + k, v)
+
+        # new executor
+        self._ctx = model.ctx[0] # only support single device for now
+        self._sym = model.symbol
+        self._arg_params = {k : v.as_in_context(self._ctx) for k, v in model.arg_params.items()}
+        self._aux_params = {k : v.as_in_context(self._ctx) for k, v in model.aux_params.items()}
+        self._executor_with_feature = False
+
+    def _init_executor(self, sym):
+        #self._arg_params["data"] = _ndarray.zeros(self._input_shape, self._ctx)
+        self._executor = sym.bind(self._ctx, self._arg_params, args_grad=None,
+                                  grad_req="null", aux_states=self._aux_params)
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def labels(self):
+        return self._labels
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def version(self):
+        return self._version
+
+    def _detect(self, im):
+        im_tensor, im_info = self._preprocess(im)
+        self._arg_params["im_info"][:] = im_info
+        self._arg_params["data"] = _ndarray.array(im_tensor, ctx=self._ctx)
+        #self._executor = self._executor.reshape(data=im_tensor.shape, im_info=(1, 3))
+        self._init_executor(self._sym)
+        self._executor.forward()
+        blobs = dict(zip(self._sym.list_outputs(), [x.asnumpy() for x in self._executor.outputs]))
+        scores = blobs["cls_prob_output"]
+        bbox_deltas = blobs["bbox_pred_output"]
+        rois = blobs["rois_output"]
+        boxes = rois[:, 1:5] / im_info[0][2]
+        boxes = bbox_transform_inv(boxes, bbox_deltas)
+        boxes = clip_boxes(boxes, (im_info[0][0] / im_info[0][2],
+                                   im_info[0][1] / im_info[0][2],
+                                   3))
+        blobs["bbox_pred"] = boxes
+        del blobs["rois_output"]
+        del blobs["bbox_pred_output"]
+        return blobs
+
+
+    def _preprocess(self, im, target_size=600, max_size=1000):
+        import PIL
+        import numpy as np
+        import StringIO as _StringIO
+        # build PIL image from sframe
+        im = PIL.Image.open(_StringIO.StringIO(im._image_data))
```

Comment:
gl image can be an encoded image then im._image_data will not work. You may want to use the to_pil method()